GVFS/GVFS.Common/FileBasedCollections/BinaryPlaceholderListDatabase.cs

```diff
@@ -0,0 +1,358 @@
+            {
+                throw new FileBasedCollectionException(e);
+            }
+        }
+
+        public void WriteAllEntriesAndFlush(IEnumerable<PlaceholderEvent> updatedPlaceholders)
+        {
+            try
+            {
+                this.WriteAndReplaceDataFile(() => this.GenerateDataLines(updatedPlaceholders));
+            }
+            catch (Exception e)
+            {
+                throw new FileBasedCollectionException(e);
+            }
+        }
+
+        private static void SerializeEvent(BinaryWriter writer, PlaceholderEvent placeholderEvent)
+        {
+            placeholderEvent.Serialize(writer);
+        }
+
+        private IEnumerable<Tuple<byte, PlaceholderEvent>> GenerateDataLines(IEnumerable<PlaceholderEvent> updatedPlaceholders)
+        {
+            HashSet<string> keys = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            this.EstimatedCount = 0;
+            foreach (PlaceholderEvent updated in updatedPlaceholders)
+            {
+                if (keys.Add(updated.Path))
+                {
+                    this.EstimatedCount++;
+                }
+
+                yield return Tuple.Create(AddEntryPrefix, updated);
+            }
+
+            if (this.placeholderChangesWhileRebuildingList != null)
+            {
+                foreach (Tuple<byte, PlaceholderEvent> entry in this.placeholderChangesWhileRebuildingList)
+                {
+                    if (entry.Item1 == RemoveEntryPrefix)
+                    {
+                        if (keys.Remove(entry.Item2.Path))
+                        {
+                            this.EstimatedCount--;
+                            yield return entry;
+                        }
+                    }
+                    else
+                    {
+                        if (keys.Add(entry.Item2.Path))
+                        {
+                            this.EstimatedCount++;
+                        }
+
+                        yield return entry;
+                    }
+                }
+
+                this.placeholderChangesWhileRebuildingList = null;
+            }
+        }
+
+        private void AddAndFlush(PlaceholderEvent entry)
+        {
+            try
+            {
+                this.WriteAddEntry(
+                    entry,
+                    () =>
+                    {
+                        this.EstimatedCount++;
+                        if (this.placeholderChangesWhileRebuildingList != null)
+                        {
+                            this.placeholderChangesWhileRebuildingList.Add(Tuple.Create(AddEntryPrefix, entry));
+                        }
+                    });
+            }
+            catch (Exception e)
+            {
+                throw new FileBasedCollectionException(e);
+            }
+        }
+
+        private bool TryParseAddLine(BinaryReader reader, out string key, out PlaceholderEvent value, out string error)

```

Comment:
This ties us tightly to the underlying format. Is this not something we can put into the collection so we don't need to be aware of the format here?